### PR TITLE
Docs: Add/update resource handler topics for datasource and app plugins

### LINKED ISSUE #878: Docs: Modify new resource handlers topic to apply to app plugins
New resource handler topic could be modified to apply to app plugins.   Suggested by Marcus in https://github.com/grafana/plugin-tools/pull/824#discussion_r1552026295:

"This applies to both data source and app plugins with a backend so there's a nice opportunity to make it cover both. Up to you whether you want to do it in this PR or follow up later.

This example cover using resource handlers in an app: [grafana/grafana-plugin-examples@main/examples/app-with-backend](https://github.com/grafana/grafana-plugin-examples/tree/main/examples/app-with-backend?rgh-link-date=2024-04-04T16%3A17%3A59Z)"

### LINKED ISSUE #1032: Docs: Add a doc about using MUX with backend plugins (app and datasources)

### PR DESCRIPTION
**Pull request by marefr**
Closes https://github.com/grafana/plugin-tools/issues/878
Closes https://github.com/grafana/plugin-tools/issues/1032

### CODE REVIEW

File: docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md [hunk 1]
```diff
@@ -47,20 +47,27 @@ To verify the data source UID, you can enter `window.grafanaBootData.settings.da
 
 ## Add support for multiple routes
 
-To support multiple routes in your data source plugin, you can use a switch with the `req.Path`:
+To support multiple routes in your data source plugin you have a couple of options depending on requirements and needs.
```
Comment (andresmgot) [inline docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md:50]: I guess all this is duplicated now so better to link to the new guide?

**Review by andresmgot — COMMENTED** [review]
Content LGTM!

File: docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md (hunk 1 above)
Comment (marefr) [reply docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md:50]: Thanks. I'm not sure I follow. Do you mean link from add resource handler topic for apps to add resource handler topic for datasources?

Comment (andresmgot) [reply docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md:50]: yep, the content is mostly (if not all) the same, isn't it?

Comment (marefr) [reply docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md:50]: Yes true, but not sure what's wanted. The main problem/missing thing is that we cannot put topics specifically for backend plugins somewhere currently. I've been trying to vote for such a section, but seems like we haven't found a good fit for that yet. I'll check with the rest of the team, thanks

Comment (academo) [reply docusaurus/docs/how-to-guides/data-source-plugins/add-resource-handler.md:50]: @marefr in the meantime, what if you create a shared file and import it here? that way it looks like we have two pages with the same information but they come from the same source.

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  - resource handler
+---
+
+# Add resource handler for app plugins
+
+You can add a resource handler to your app backend to extend the Grafana HTTP API with your own app-specific routes. This guide explains why you may want to add [resource](../../key-concepts/backend-plugins/#resources) handlers and some common ways for doing so.
+
+## Uses of resource handlers
+
+The primary way for a data source to retrieve data from a backend is through the [query method](../../tutorials/build-a-data-source-plugin#define-a-query). But sometimes your data source needs to request data on demand; for example, to offer auto-completion automatically inside the data source’s query editor.
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:20]: this paragraphs talks about data source but this is an app guide

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+		})
+	}
+}
+```
+
+Supporting additional routes and methods (GET, POST etc) pretty quickly gets cumbersome and creates code that might be hard to maintain, test and read. Here we recommend to use the more Go-agnostic approach for handling resources and use the regular [`http.Handler`](https://pkg.go.dev/net/http#Handler). You can do so by using a package provided by the [Grafana Plugin SDK for Go](../../key-concepts/backend-plugins/grafana-plugin-sdk-for-go) named [`httpadapter`](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter). This package provides support for handling resource calls using an [`http.Handler`](https://pkg.go.dev/net/http#Handler).
+
+Using [`http.Handler`](https://pkg.go.dev/net/http#Handler) allows you to also use Go’s built-in router functionality called [`ServeMux`](https://pkg.go.dev/net/http#ServeMux) or your preferred HTTP router library (for example, [`gorilla/mux`](https://github.com/gorilla/mux)).
+
+:::note
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:77]: It'd much prefer this was the first content in the article instead of a side note. Using the ServeMux should be the "default" way to go about calling resources in  most cases.

What if move this on top explaining in the section `Implement the resource handler interface` that you have two options: Use ServeMux or manually implement the interface. Then you explain how to do it with ServeMux and finally how to do it using the call resource handler. e.g.:
---

# Implement the resource handler interface

To add a resource handler to your app plugin, you need to implement the `backend.CallResourceHandler` interface for your app struct.

There are two ways you can implement this in your plugin, using ServeMux or manually implementing it in your plugin:

## Using ServeMux

// this whole section

## Manually implementing `backend.CallResourceHandler` 

// the original part with and the multiple routes section go here

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+		return
+	}
+
+	rw.WriteHeader(http.StatusOK)
+}
+```
+
+## Accessing app resources from the frontend
+
+You can query your resources using the `getResource` and `postResource` helpers from the `DataSourceWithBackend` class. To provide a nicer and more convenient API for your components it's recommended to extend your datasource class and instance with functions for each route as shown in the following example:
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:224]: This content is about data sources, but this guide is about app plugins so it won't work.

**Review by academo — COMMENTED** [review]
You should attempt to make the content about CallResourceHandler generic (no mention of app or data source), put it in a shared file and import it in both guides. Then in data sources you can put the specific frontend code and in app the specific frontend app code.

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+With the above example you can access the following resources:
+
+- HTTP GET `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/namespaces`
+- HTTP GET `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/projects`
+- HTTP POST `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/device`
+
+### From the frontend
+
+You can access your resources using the `get` and `post` functions from the `backendSrv` runtime service. To provide a nicer and more convenient API for your components it's recommended to provide a helper class with functions for each route as shown in the following example:
```
Comment (marefr) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:49]: I could use some help with the examples here, not sure they're correct/what we want to encourage?

**Comment by marefr** [comment]
> You should attempt to make the content about CallResourceHandler generic (no mention of app or data source), put it in a shared file and import it in both guides. Then in data sources you can put the specific frontend code and in app the specific frontend app code.

@academo @andresmgot I think I've managed to do what you suggested. PTAL 🙏

File: docusaurus/docs/shared/implement-resource-handler.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  ctxLogger := backend.Logger.FromContext(req.Context())
+
+  rw.Header().Add("Content-Type", "application/json")
+  _, err := rw.Write([]byte(`{ "namespaces": ["ns-1", "ns-2"] }`))
+  if err != nil {
+    ctxLogger.Error("Failed to write response", "error", err)
+    return
+  }
+  rw.WriteHeader(http.StatusOK)
+}
```
Comment (andresmgot) [inline docusaurus/docs/shared/implement-resource-handler.md:65]: for the sake of simplicity maybe we can have a shorter version:

```suggestion
func (p *MyPlugin) handleNamespaces(rw http.ResponseWriter, req *http.Request) {
  rw.Header().Add("Content-Type", "application/json")
  _, err := rw.Write([]byte(`{ "namespaces": ["ns-1", "ns-2"] }`))
  if err != nil {
    rw.WriteHeader(http.StatusInternalServerError)
    return
  }
  rw.WriteHeader(http.StatusOK)
}
```

Up to you if you want to leave it as it is since it's more complete.

File: docusaurus/docs/shared/implement-resource-handler.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    return
+  }
+
+  rw.WriteHeader(http.StatusOK)
+}
+```
+
+#### Accessing the backend plugin context
+
+Use the `backend.PluginConfigFromContext` function to access `backend.PluginContext`:
```
Comment (andresmgot) [inline docusaurus/docs/shared/implement-resource-handler.md:134]:
```suggestion
Additionally, you can use the `backend.PluginConfigFromContext` function to access `backend.PluginContext`. This [struct](https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.247.0/backend/common.go#L166) contain information that may be useful, like the user performing the request:
```

**Review by andresmgot — APPROVED** [review]
LGTM, I only have a couple minor suggestions

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md (hunk 5 above)
Comment (academo) [reply docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:49]: Using an "API" class it not something common in frontend development, it is more common to simply have a module with functions inside. 

For this example I simply using `fetch` directly in the places where you are now using `api.` and remove all the abstraction (i'll add comments in those places)

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    return this.backend.post(`/api/plugins/${this.PluginID}/resources/device`, { state: state });
+  }
+}
+```
+
+For example, in your app or component you can instantiate your API class and use `getNamespaces` to send a HTTP GET request to `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/namespaces`
+
+```typescript
+const api = new API('my-app-id');
+const namespaces = await api.getNamespaces();
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:77]: Let's remove the API class abstraction and simply use getBackendSrv directly. let developers figure their own pattern. 
```suggestion
const namespaces = await getBackendSrv().get(`/api/plugins/${pluginID}/resources/namespaces`);
```

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+import ImplementResourceHandler from '@shared/implement-resource-handler.md';
+
+# Add resource handler for app plugins
+
+You can add a resource handler to your app backend to extend the Grafana HTTP API with your own app-specific routes. This guide explains why you may want to add [resource](../../key-concepts/backend-plugins/#resources) handlers and some common ways for doing so.
+
+## Uses of resource handlers
+
+The use case and functionality for an app is very broad and therefore also for uses of resource handlers. But in general, an app normally integrates with a HTTP service of some kind, e.g. a 3rd party service, to retrieve and send data. For example, this service might have
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:22]: `The use case and functionality for an app is very broad and therefore also for uses of resource handlers.`

This sentence confused me a bit, could you rephrase it?

File: docusaurus/docs/shared/implement-resource-handler.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  rw.Header().Add("Content-Type", "application/json")
+  _, err := rw.Write([]byte(`{ "namespaces": ["ns-1", "ns-2"] }`))
+  if err != nil {
+    ctxLogger.Error("Failed to write response", "error", err)
+    return
+  }
+  rw.WriteHeader(http.StatusOK)
+}
+
+func (p *MyPlugin) handleProjects(rw http.ResponseWriter, req *http.Request) {
```
Comment (academo) [inline docusaurus/docs/shared/implement-resource-handler.md:67]: IMO there's no need to provide fully functional code snippets for these resource handlers. create-plugin already generates functional code that devs can see working in their code. 

perhaps you can reduce this to simply the response removing all the code in the middle? it can become distracting from the important part which is the resource handler part.

File: docusaurus/docs/shared/implement-resource-handler.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+```
+
+#### Accessing the backend plugin context
+
+Use the `backend.PluginConfigFromContext` function to access `backend.PluginContext`:
+
+```go
+func (p *MyPlugin) handleNamespaces(rw http.ResponseWriter, req *http.Request) {
+	pCtx := backend.PluginConfigFromContext(req.Context())
+	ctxLogger := backend.Logger.FromContext(req.Context())
```
Comment (academo) [inline docusaurus/docs/shared/implement-resource-handler.md:139]: it'd be nice to have an example here of how to access the plugin configuration from this handler using the context.

Comment (marefr) [reply docusaurus/docs/shared/implement-resource-handler.md:139]: I think that's a separate topic on its own so I'll create a new issue for that

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md (hunk 5 above)
Comment (marefr) [reply docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md]: I've simplified it

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  updateDevice(state: string): Promise<string> {
+    return this.backend.post(`/api/plugins/${this.PluginID}/resources/device`, { state: state });
+  }
+}
+```
+
+For example, in your app or component you can instantiate your API class and use `getNamespaces` to send a HTTP GET request to `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/namespaces`
+
+```typescript
+const api = new API('my-app-id');
```
[the remark this answers is not in the document]
Comment (marefr) [reply docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md]: Skipping that for now to simplify

File: docusaurus/docs/shared/implement-resource-handler.md (hunk 10 above)
Comment (academo) [reply docusaurus/docs/shared/implement-resource-handler.md:58]: Much better. thanks <3

File: docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+- HTTP GET `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/projects`
+
+### From the frontend
+
+You can access your resources in a compontent using the `get` function of the `backendSrv` runtime service to send a HTTP GET request to `http://<GRAFANA_HOSTNAME>:<PORT>/api/plugins/<PLUGIN_ID>/resources/namespaces`
+
+```typescript
+import { getBackendSrv } from '@grafana/runtime';
+
+const namespaces = await getBackendSrv().get(`/api/plugins/<PLUGIN_ID>/resources/namespaces`);
```
Comment (academo) [inline docusaurus/docs/how-to-guides/app-plugins/add-resource-handler.md:50]: much nicer! thanks :D

**Review by academo — APPROVED** [review]

File: docusaurus/docs/shared/implement-resource-handler.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  p := &MyPlugin{}
+  mux := http.NewServeMux()
+  mux.HandleFunc("/namespaces", p.handleNamespaces)
+  mux.HandleFunc("/projects", p.handleProjects)
+  p.resourceHandler := httpadapter.New(mux)
+  return p
+}
+
+func (p *MyPlugin) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
+  return p.resourceHandler.CallResource(ctx, req)
```
Comment (andresmgot) [inline docusaurus/docs/shared/implement-resource-handler.md:46]: FTR, I just noticed that this had a typo, it should be:

```
  return p.resourceHandler.CallResource(ctx, req, sender)
```

Comment (marefr) [reply docusaurus/docs/shared/implement-resource-handler.md:46]: Thanks. Let me fix
